okhttp/src/main/java/okhttp3/CipherSuite.java

```diff
@@ -29,356 +32,421 @@
  * <p>See also <a href="https://android.googlesource.com/platform/external/conscrypt/+/master/src/main/java/org/conscrypt/NativeCrypto.java">NativeCrypto.java</a>
  * from conscrypt, which lists the cipher suites supported by Android.
  */
-public enum CipherSuite {
+public final class CipherSuite {
+  // for interning - must be at top of file
+  private static final ConcurrentMap<String, CipherSuite> INSTANCES = new ConcurrentHashMap<>();
```

Comment:
I actually changed this now to use a WeakHashMap guarded by a ReentrantReadWriteLock.

The reason I'd like weak references is that the values from which we construct CipherSuite instances may come from a (potentially misbehaving) server, so keeping any value ever mentioned in strongly referenced memory may not be a good idea. The new code only keeps those CipherSuite instances around that are either strongly referenced themselves or whose javaName is strongly referenced (e.g. those that occur as String literals). I've added a test for this as well which calls Runtime.gc().

Re copy-on-write LinkedHashMap: Why copy-on-write? This map is actually modified many times during <clinit> so this copy-on-write would make this O(n^2) without additional care/boilerplate. Re the "Linked" part, why Linked? This Map is never iterated over so the iteration order shouldn't matter.

Let me know if you have any concerns about my "WeakHashMap guarded by a ReentrantReadWriteLock" approach.